### PR TITLE
bpo-35810: Incref heap-allocated types in PyObject_Init

### PR DESCRIPTION
**Pull request by eduardo-elizondo**
https://bugs.python.org/issue35810

Correctly incref an intance's type. Currently, if a heap-allocated type creates an instance through PyObject_{,GC}_New{Var}, the type won't incref. This adds a change to pull the incref to the core PyObject_Init{INIT} function to correctly incref heap-allocated types.

This now means that heap-allocated types that add a custom tp_dealloc, should decref the instance types - just like the default tp_dealloc does.

Currently there are 10 heap-allocated types in CPython:
* PyCursesPanel_Type in _curses_panel.c
* sslerror_type in _ssl.c
* Example_Type in _testmultiphase.c
* Str_Type in _testmultiphase.c
* Tkapp_Type in _tkinter.c
* Tktt_Type in _tkinter.c
* PyTclObject_Type in _tkinter.c
* Xxo_Type in xxlimited.c
* Str_Type in xxlimited.c
* Null_Type in xxlimited.c
* Struct Sequences in structseq.c

Out of those only the following 5 types allocate instances through PyObject_{GC}_New{Var}:
* PyCursesPanel_Type. Action: Added a decref in its tp_dealloc
* Tkapp_Type. Action: Removed the manual incref after allocation.
* Tktt_Type. Action: Removed the manual incref after allocation.
* PyTclObject_Type. Action: Removed the manual incref after allocation.
* Xxo_Type. No Action: It inherits the default tp_dealloc which already decrefs the type.
* Struct Sequences. Action: Removed decref the type in the deallocation function.

### CODE REVIEW

**Comment by eduardo-elizondo** [comment]
Previous reference: https://github.com/python/cpython/pull/10854#discussion_r250284027

cc @vstinner

**Review by vstinner — DISMISSED** [review]
Sorry but this change looks the obviously break the backward compatibility on all heap-type used in the wild. I don't think that it's a good idea to break the backward compatibility like that. Or maybe I misunderstood something.

cc @encukou

**Comment by vstinner** [comment]
cc @pablogsal @serhiy-storchaka @scoder

**Comment by eduardo-elizondo** [comment]
I have made the requested changes; please review again

Hi @vstinner, is there any particular scenario that you are concerned about? 

This is indeed backwards compatible. The single case where behavior is not preserved causes a type to become immortal. This only arises under the following set of conditions: 1) A user manually jams `NULL` into the `tp_new` slot of a heap-allocated type. 2) The user manually increfs the type after object instantiation, and 3) The `tp_dealloc` slot decrefs the type. Outside of that, behavior is preserved. This is a very rare set of circumstances which most likely do not exist in the wild.

Anyways, let me know what you think. If you'd like to, I'll be happy to follow-up with a more thorough explanation on the correctness of this change.

**Comment by eduardo-elizondo** [comment]
cc @encukou

**Comment by encukou** [comment]
Please add an entry in `Doc/whatsnew/3.8.rst` (Build and C API Changes). This is something we'll definitely want extension authors to know about.

I do believe this is a good change to make, but we need to be very careful with it.

I ran some refleak tests (`-R :`) on Linux and found leaks in `posix.stat_result`. Reproducer:
```
>>> import os, sys, gc
>>> sys.getrefcount(os.stat_result)
149
>>> os.stat('/')
...
>>> gc.collect()
62
>>> sys.getrefcount(os.stat_result)
150
```
I assume this happens with all `PyStructSequence` types. 
If you have time to test PyStructSequence C API, that would be great (but it's a bit out of scope of this issue/PR).

It's good that the refleak tests found a failure. I'm less happy about the fact that we'll need to run them to find OS-specific failures. Not sure if our CI will cover enough ground here. (@vstinner, do you know?)

I'm almost out of time today but keeping this on my agenda.

**Comment by eduardo-elizondo** [comment]
I have made the requested changes; please review again

Woops, I was originally only checking `cpython/Modules` for heap-allocated types! I forgot that I recently modified `cpython/Objects/structseq.c` to also use `PyType_FromSpec`. The fix is just to have structseq to decref the Type in `tp_dealloc`. It's actually covered under the 4 scenarios that discussed here: https://github.com/python/cpython/pull/10854#issuecomment-455744745. Refleak is happy now 👍 

whatsnew/3.8.rst has been updated as well.

--------------------

On another note, @encukou if you want to, on another PR, I can add more documentation on moving types from `PyType_Ready` to `PyType_FromSpec`. There are different scenarios that might lead to leaks and/or crashes. I've already documented some of the scenarios in the comment that I just pointed you to but maybe it should be moved to somewhere more discoverable. I think this will greatly reduce the debugging time for external users. Just let me know.

**Comment by encukou** [comment]
> On another note, @encukou if you want to, on another PR, I can add more documentation on moving types from `PyType_Ready` to `PyType_FromSpec`. There are different scenarios that might lead to leaks and/or crashes. I've already documented some of the scenarios in the comment that I just pointed you to but maybe it should be moved to somewhere more discoverable. I think this will greatly reduce the debugging time for extension authors. Just let me know.

That sounds great! Two things I'd expect from a HOWTO for the documentation:
* It should say why heap types are better, and in which cases, not just that everyone should switch
* Let's not document bugs as features: if there's something wrong in the current implementation, track it in issues instead :)

But even notes about the current state would be useful.

I will get to this next week again. Thank you very much for the effort, and I hope my limited time is not too discouraging.

**Comment by eduardo-elizondo** [comment]
> That sounds great! Two things I'd expect from a HOWTO for the documentation:
Perfect! I'll get something nice written up this week!

> I hope my limited time is not too discouraging.
Not at all! Really appreciate the reviews 😄.

Looking forward to see what you think about moving this forward!

**Comment by vstinner** [comment]
This change is going to break basically all C extensions which implement types. I would prefer to see a discussion on python-dev to make sure that we want to break the backward compatibility on purpose.

The strict minimum would be to test a few C extensions to check how many are broken by your PR. Examples: numpy, pandas, Cython, Pillow, lxml, etc.

**Comment by vstinner** [comment]
@scoder: Would you mind to review this change?

**Comment by vstinner** [comment]
I'm not strictly against "fixing" the C API. But any backward incompatible change must be **very carefully* prepared and organized: communicate with maintainers of major C extensions, communicate properly on the change, maybe provide tooling to detect bugs and/or do the migration.

**Comment by eduardo-elizondo** [comment]
It seems that all the issues have already been addressed as well as adding more documentation for porting into 3.8.

Let me know if there are any other pending changes.

cc @vstinner

**Comment by eduardo-elizondo** [comment]
I have made the requested changes; please review again

I seems that everyone is in agreement about this change in https://bugs.python.org/issue35810. Let's try to push this change forward! cc @vstinner @encukou

**Comment by eduardo-elizondo** [comment]
Turns out that using the builtin github merger is a pain 😞 . Fixing

**Comment by encukou** [comment]
Hi!
I hoipe you don't mind pushing to the PR directly. I reworded the release notes documentation to hopefully make them clearer – it's IMO more efficient that way than giving you a list of nitpicks to sort out.
Let me know what you think.

**Comment by eduardo-elizondo** [comment]
@encukou looks great, it reads much better now. Thank you! 👍

**Comment by encukou** [comment]
- *discussion on python-dev*: [happened](https://mail.python.org/pipermail/python-dev/2019-February/156322.html), discussion was directed to the issue.
   - *test a few C extensions to check how many are broken*: 
      - numpy: a few unrelated tests fail (even with CPython master):
        - `TestPickling.test_highest_available_pickle_protocol` (numpy assumes 3.8 has PEP 574; we are not there yet)
        - `TestRandomDist.test_binomial` (deprecation warning treated as error)
        - `TestRandomDist.test_hypergeometric` (deprecation warning treated as error)
        - `TestRandomDist.test_scalar_exception_propagation` (related to hypergeometric)
      - pandas: tests coredump for me (even with CPython master) :(
      - Cython: build of dependencies fails on 3.8 PyThreadState changes (even with CPython master)
      - Pillow: tests pass
      - lxml: tests pass
   - *communicate with maintainers of major C extensions*: As it turns out the major C extensions do not use heap-allocated types (yet). AFAICS:
      - Numpy doesn't.
      - Pandas doesn't
      - Cython doesn't (but Stefan is on board here).
      - Pillow doesn't
      - lxml doesn't
      - PySide does (and Christian is aware).
      Anything more?
   - *communicate properly on the change*: I think the What's New entries are now adequate.
   - *maybe provide tooling to detect bugs and/or do the migration* - while tooling to detect refcount bugs would definitely be great, this sounds like overkill to me.
     -  Extension authors should really be running refleak checks *now*. They mostly aren't, and that's partly on us for not providing the tools, but I think that's out of scope for this issue.
     - The migration is not difficult (see the Wat's New entry), and rare (heap types aren't widely used, and the brave early adopters who do use them seem prepared for some ironing out of issues).

@vstinner, do you have any more reservations?

**Comment by scoder** [comment]
> Cython: build of dependencies fails on 3.8 PyThreadState changes (even with CPython master)

Note sure what you mean here. Cython doesn't have any dependencies. Its test suite uses some libraries when they are installed (line profiler, numpy, coverage), but without them it simply disables a couple of tests and that's it.

**Comment by scoder** [comment]
I just noticed that I forgot to express a big "Thank You!" to @encukou for digging up the references above and providing an excellent ground for estimating the risks.

FWIW, I ran Cython's test suite with this branch and it didn't show any issues. This doesn't guarantee that there won't be issues in third party Cython code, but at least (as expected) Cython itself doesn't seem to produce any problems here, which makes it very unlikely that many Cython implemented packages will suffer from this change. People sometimes do hackish C stuff in their Cython code, but probably not in this particular corner.

File: Doc/whatsnew/3.8.rst
Comment (scoder) [inline Doc/whatsnew/3.8.rst:706]: In Cython, we normally use the exact alpha/beta version where the change was introduced (to support testing during the CPython release cycle), but we obviously don't know that yet.

File: Objects/object.c
Comment (scoder) [inline Objects/object.c:246]: Shouldn't this just call `_PyObject_INIT()` ?

**Review by scoder — APPROVED** [review]

File: Objects/object.c
Comment (scoder) [reply Objects/object.c:246]: Or, actually, shouldn't `PyObject_Init()` just call `_PyObject_INIT()` ? Duplicating that code across two files seems unnecessary, maybe even a bit error prone.

Comment (encukou) [reply Objects/object.c:246]: `PyObject_Init` is defined in the same file, so the compiler should have everything it needs to inline as it sees fit.

As for code duplication, that's a valid point, but I'd call that out of scope for this PR. Let's keep orthogonal issues revertable separately.

File: Doc/whatsnew/3.8.rst
Comment (encukou) [reply Doc/whatsnew/3.8.rst:706]: The example is not really written for projects with Cython's level of involvement. Most projects will only test against a nightly build or the latest alpha/beta (if even that). The few others should know to adjust the example. We don't explicitly tell you to rename `foo_new`, either :)

**Comment by eduardo-elizondo** [comment]
Would it make sense to request a review on python-dev? Otherwise, @encukou how would you feel about reviewing the PR?

**Comment by encukou** [comment]
Sorry for the delay! I missed the notification about Victor dismissing his review :(

I am available today to watch the buildbots, so I'll merge.

**Pull request merged**
